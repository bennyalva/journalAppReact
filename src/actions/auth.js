import { types } from "../types/types"
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { startLoading, finishLoading } from "./ui";
import Swal from 'sweetalert2'
import { noteLogout } from "./notes";

export const startLoginEmailPassWord = (email, password) => {
    
    return (dispatch) => {
            dispatch(startLoading()); // este dispatch es el unico que se regresa cuando se hace testitng
            // para que los otros dipatch se muestren en el testing se agrega un return en firebase
            firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ({user}) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
             }).catch(e => {
                dispatch(finishLoading());
                Swal.fire('Error', e.message, 'error')
              });
    }
}

export const startReigisteredUser = (email, password, name) =>{
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then( async ({user}) => {
            await user.updateProfile({ displayName: name});
            dispatch(login(user.uid, user.displayName)) 
         }).catch(e => {
            Swal.fire('Error', e.message, 'error');
    
         });
    }
}
export const startGoogleLogin = () => {
    return (dispatch) => { /// cuando hacemos el return y va dentro los dispacth se puede ver cuando se hacemos testing
        firebase.auth().signInWithPopup( googleAuthProvider)
        .then( ({user}) => {
           dispatch(login(user.uid, user.displayName)) 
        });
    }
}
export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
    });

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(noteLogout());
    }
} 

export const logout = () => (
    {
        type: types.logout
    }
);