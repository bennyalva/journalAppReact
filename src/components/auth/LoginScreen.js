import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux'
import { startLoginEmailPassWord, startGoogleLogin } from '../../actions/auth'

export const LoginScreen = () => {
    const { loading } = useSelector(state => state.ui);
    const dispatch = useDispatch();
    const [formValues, handlerInputChanges] = useForm({
        email: 'benny@gmail.com',
        password: '123456'
    });

    const { email, password} = formValues;

    const handlerLogin = (e) => {
        e.preventDefault();
        dispatch(startLoginEmailPassWord(email, password));
    }
    const handlerGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }
    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form onSubmit={ handlerLogin } 
            className="animate__animated animate__fadeIn animate_faster">
                <input className="auth__input"
                    type="text"
                    placeholder="Email"
                    name="email"
                    autoComplete="off"
                    value={ email }
                    onChange={ handlerInputChanges }
                />
                <input
                    className="auth__input"
                    type="Password"
                    placeholder="password"
                    name="password"
                    value={password}
                    onChange={ handlerInputChanges }
                />
                <button
                    className="btn btn-primary btn-block"
                    type="submit"
                    disabled={ loading }>
                    Login
                </button>
                <hr />
                <div  className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div 
                            className="google-btn"
                            onClick={handlerGoogleLogin}
                        >
                            <div className="google-icon-wrapper">
                                <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                            </div>
                            <p className="btn-text">
                                <b>Sign in with google</b>
                            </p>
                        </div>
                </div>
                <Link to="/auth/register" className="link">
                  Create account
                </Link>
            </form>
        </>
    )
}
