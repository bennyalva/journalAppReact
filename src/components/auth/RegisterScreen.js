import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { setErrorAction, removeErrorAction } from '../../actions/ui';
import { startReigisteredUser } from '../../actions/auth';

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError } = useSelector(state => state.ui);


    const [formValues, handlerInputChanges] = useForm({
        name: 'benny',
        email: 'benny@gmail.com',
        password: '123456',
        passeord2: '123456'
    });
    const { name, email, password , passeord2} = formValues;

    const handlerSubmitRegister = (e) => {
        e.preventDefault();
        if ( isFormValid()) {
            dispatch(startReigisteredUser(email, password, name));
        }
    }

    const isFormValid = () => {
      if (name.trim().length === 0) {
          dispatch(setErrorAction('name is crequired'));
        return false;
      } else if ( !validator.isEmail( email )) {
        dispatch(setErrorAction('Email is not valid'));
        return false;
      } else if ( password !== passeord2 || password.length < 5) {
        dispatch(setErrorAction('Password should be at least 5 cracters and match aeach other'));
          return false;
      }
      dispatch(removeErrorAction());
      return true;
    }
    return (
        <>
        <h3 className="auth__title">Register</h3>
        <form onSubmit={ handlerSubmitRegister }>
            {
                msgError && (
                   <div className="auth__alert-error">
                       {msgError}
                   </div>
                )
            }
            <input className="auth__input"
                type="text"
                placeholder="Name"
                name="name"
                autoComplete="off"
                value={ name }
                onChange={ handlerInputChanges }
            />
            <input className="auth__input"
                type="text"
                placeholder="Email"
                name="email"
                autoComplete="off"
                value={ email }
                onChange={ handlerInputChanges }
            />
            <input
                className="auth__input"
                type="Password"
                placeholder="password"
                name="password"
                value={ password }
                onChange={ handlerInputChanges }
            />
            <input className="auth__input"
                type="text"
                placeholder="Confirm password"
                name="passeord2"
                autoComplete="off"
                value={ passeord2 }
                onChange={ handlerInputChanges }
            />
            <button
                className="btn btn-primary btn-block mb-1"
                type="submit">
                Register
            </button>
    
            
            <Link to="/auth/login" className="link">
              Already register ?
            </Link>
        </form>
    </>
    )
}
