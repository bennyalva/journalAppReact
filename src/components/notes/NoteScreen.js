import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useSelector, useDispatch } from 'react-redux'
import { useForm } from '../../hooks/useForm';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {
    const { active: note} = useSelector(state => state.notes);
    const [formValues, handlerInputChanges, reset] = useForm( note );
    const { body, title, id} = formValues;

    const dispatch = useDispatch();

    const activeId = useRef( note.id)
    useEffect(() => {
        if ( note.id !== activeId.current ) {
            reset( note );
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
       dispatch(activeNote(formValues.id, { ...formValues }));
    }, [formValues, dispatch]);

    const handlerDelete = () => {
        dispatch(startDeleting( id))
    }
    return (
        <div className="notes__main-content">
            <NotesAppBar />
            <div className="notes__content">
                <input
                        type="text"
                        autoComplete="off"
                        placeholder="Some awesome title"
                        className="notes__title-input"
                        value={ title }
                        name="title"
                        onChange={ handlerInputChanges } />
                
                <textarea 
                        placeholder="What hapeend today ..."
                        className="note__textarea"
                        value={ body }
                        name="body"
                        onChange={ handlerInputChanges }
                >

                </textarea>
                {
                    (note.url)&& <div className="note__images">
                    <img src={note.url} alt="imagen" />
                </div>
                }
            </div>
            <button className="btn btn-danger"
                    onClick={ handlerDelete}>
                Delete
            </button>
        </div>
    )
}
