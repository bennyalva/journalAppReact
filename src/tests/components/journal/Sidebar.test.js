import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';
const { mount } = require("enzyme");
const { Sidebar } = require("../../../components/journal/Sidebar");

jest.mock('../../../actions/auth', () => (
    {
        startLogout: jest.fn()
    }
));
jest.mock('../../../actions/notes', () => (
    {
        startNewNote: jest.fn()
    }
));
const middlewares = [thunk]; // config store
const mockStore = configureStore(middlewares); // config store

const initState = { // inicia es estado del store que se necesita cuando se renderiza el componente
    auth: {
        uid: 123,
        name: 'benny'
    },
    ui: {
        loading: false,
        msgError: ''
    },
    notes: {
        active: null,
        notes: [
            {
                id: 123, 
                date: 978098723, 
                title: 'title', 
                body: 'body',
                url: 'https://images'
            }
        ]
    }
};

let store  = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <Sidebar />
    </Provider>)
describe('Test on Sidebar', () => {

   test('should show correctly', () => {
     expect(wrapper).toMatchSnapshot();     
   });
   test('should call logout', () => {
       console.log(wrapper.find('button').html())
       wrapper.find('button').prop('onClick')();
       expect(startLogout).toHaveBeenCalled();
   });
   test('should call startNewnote', () => {
       wrapper.find('.journal__new-entry').prop('onClick')();
       expect(startNewNote).toHaveBeenCalled()
   });
   
   
    
});
