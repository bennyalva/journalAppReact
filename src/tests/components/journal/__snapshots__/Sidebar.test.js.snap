// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on Sidebar should show correctly 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        benny
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  >
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate_faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://images)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          title
        </p>
        <p
          className="journal__entry-content"
        >
          body
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Monday
        </span>
        <h4>
           
          12th
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
